File: paddle3d/ops/bev_feature_ad2/bev_feature.cpp
// Copyright (c) 2023 PaddlePaddle Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <vector>

#include "paddle/extension.h"

#define CHECK_INPUT_CUDA(x) \
  PD_CHECK(x.is_gpu() || x.is_gpu_pinned(), #x " must be a GPU Tensor.")

void mapfeatureLauncher(const cudaStream_t& stream, const int num,
                        const int point_dim, const float* points_data,
                        const int* points2grid_data, float* max_height_data,
                        float* mean_height_data, float* mean_intensity_data,
                        float* count_data);
void TopIntensityLauncher(const cudaStream_t& stream, const int num,
                          const int point_dim, const float* points_data,
                          const int* points2grid_data,
                          float* top_intensity_data, float* max_height_data);
void ComputeLogOddsGPULauncher(
    const cudaStream_t& stream, const int num, const int point_dim,
    const float voxel_size_x, const float voxel_size_y,
    const float voxel_size_z, const float pxmin, const float pymin,
    const float pzmin, const int grid_size_x, const int grid_size_y,
    const int grid_size_z, const float offset_origin_x,
    const float offset_origin_y, const float offset_origin_z, float lo_occupied,
    float lo_free, const float* points_data, float* logodds_data);

std::vector<paddle::Tensor> cnnseg_feature_generator_gpu(
    const paddle::Tensor& points, const paddle::Tensor& points2grid,
    const std::vector<int>& grid_size, const int cnnseg_feature_dim) {
  CHECK_INPUT_CUDA(points);
  CHECK_INPUT_CUDA(points2grid);  

  const int grid_size_x = grid_size[0];
  const int grid_size_y = grid_size[1];
  const int grid_size_z = grid_size[2];

  auto cnnseg_feature =
      paddle::full({cnnseg_feature_dim, grid_size_y, grid_size_x}, 0,
                   points.type(), paddle::GPUPlace());
  cnnseg_feature = paddle::scatter(cnnseg_feature, paddle::full({1}, 0, points2grid.type(), paddle::GPUPlace()), paddle::full({1, grid_size_y, grid_size_x}, -5.0, points.type(), paddle::GPUPlace()));

  const float* points_data = points.data<float>();
  const int* points2grid_data = points2grid.data<int>();
  float* cnnseg_feature_data = cnnseg_feature.data<float>();
  const int unit_offset = grid_size_x * grid_size_y;
  float* mean_height_data = cnnseg_feature_data + unit_offset * 0;
  float* max_height_data = cnnseg_feature_data + unit_offset * 1;
  float* count_data = cnnseg_feature_data + unit_offset * 2;
  float* top_intensity_data = cnnseg_feature_data + unit_offset * 3;
  float* mean_intensity_data = cnnseg_feature_data + unit_offset * 4;
  float* nonempty_data = cnnseg_feature_data + unit_offset * 5;
  const int num_point = points.shape()[0];
  const int point_dim = points.shape()[1];
  cudaStream_t stream = points.stream();
  mapfeatureLauncher(stream, num_point, point_dim, points_data,
                     points2grid_data, max_height_data, mean_height_data,
                     mean_intensity_data, count_data);

  TopIntensityLauncher(stream, num_point, point_dim, points_data,
                       points2grid_data, top_intensity_data, max_height_data);

  return {cnnseg_feature};
}

std::vector<paddle::Tensor> visibility_feature_gpu(
    const paddle::Tensor& points, const std::vector<float>& vis_voxel_size,
    const std::vector<float>& vis_range, const std::vector<int>& vis_grid_size,
    const std::vector<float>& vis_offset_origins, const float lo_occupied,
    const float lo_free) {
  CHECK_INPUT_CUDA(points);

  const float voxel_size_x = vis_voxel_size[0];
  const float voxel_size_y = vis_voxel_size[1];
  const float voxel_size_z = vis_voxel_size[2];
  const float pc_start_x = vis_range[0];
  const float pc_start_y = vis_range[1];
  const float pc_start_z = vis_range[2];
  const int grid_size_x = vis_grid_size[0];
  const int grid_size_y = vis_grid_size[1];
  const int grid_size_z = vis_grid_size[2];
  const float offset_origin_x = vis_offset_origins[0];
  const float offset_origin_y = vis_offset_origins[1];
  const float offset_origin_z = vis_offset_origins[2];

  auto logodds_feature =
      paddle::full({vis_grid_size[2], vis_grid_size[1], vis_grid_size[0]}, 0,
                   points.type(), paddle::GPUPlace());
  const float* points_data = points.data<float>();
  float* logodds_feature_data = logodds_feature.data<float>();
  int num_point = points.shape()[0];
  int point_dim = points.shape()[1];
  cudaStream_t stream = points.stream();
  ComputeLogOddsGPULauncher(
      stream, num_point, point_dim, voxel_size_x, voxel_size_y, voxel_size_z,
      pc_start_x, pc_start_y, pc_start_z, grid_size_x, grid_size_y, grid_size_z,
      offset_origin_x, offset_origin_y, offset_origin_z, lo_occupied, lo_free,
      points_data, logodds_feature_data);
  return {logodds_feature};
}

std::vector<std::vector<int64_t>> CnnSegFeatGenInferShape(
    std::vector<int64_t> points_shape, std::vector<int64_t> points2grid_shape,
    const std::vector<int>& grid_size, const int cnnseg_feature_dim) {
  return {{cnnseg_feature_dim, grid_size[1], grid_size[0]}};
}

std::vector<paddle::DataType> CnnSegFeatGenInferDtype(
    paddle::DataType points_dtype, paddle::DataType points2grid_dtype) {
  return {points_dtype};
}

std::vector<std::vector<int64_t>> VisFeatInferShape(
    std::vector<int64_t> points_shape, const std::vector<float>& vis_voxel_size,
    const std::vector<float>& vis_range, const std::vector<int>& vis_grid_size,
    const std::vector<float>& vis_offset_origins, const float lo_occupied,
    const float lo_free) {
  return {{vis_grid_size[2], vis_grid_size[1], vis_grid_size[0]}};
}

std::vector<paddle::DataType> VisFeatInferDtype(paddle::DataType points_dtype) {
  return {points_dtype};
}

PD_BUILD_OP(cnnseg_feature_generator_gpu)
    .Inputs({"points", "points2grid"})
    .Outputs({"cnnseg_feature"})
    .SetKernelFn(PD_KERNEL(cnnseg_feature_generator_gpu))
    .Attrs({"grid_size: std::vector<int>", "cnnseg_feature_dim: int"})
    .SetInferShapeFn(PD_INFER_SHAPE(CnnSegFeatGenInferShape))
    .SetInferDtypeFn(PD_INFER_DTYPE(CnnSegFeatGenInferDtype));

PD_BUILD_OP(visibility_feature_gpu)
    .Inputs({"points"})
    .Outputs({"logodds_feature"})
    .SetKernelFn(PD_KERNEL(visibility_feature_gpu))
    .Attrs({"vis_voxel_size: std::vector<float>",
            "vis_range: std::vector<float>", "vis_grid_size: std::vector<int>",
            "vis_offset_origins: std::vector<float>", "lo_occupied: float",
            "lo_free: float"})
    .SetInferShapeFn(PD_INFER_SHAPE(VisFeatInferShape))
    .SetInferDtypeFn(PD_INFER_DTYPE(VisFeatInferDtype));

File: paddle3d/utils_idg/ops/bev_feature.py
# !/usr/bin/env python3
"""
Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""
import paddle
from paddle import nn
from paddle3d.ops import bev_feature_layer

import numpy as np

class BevFeature(object):
    """ as name """
    def __init__(self,
                voxel_size,
                point_cloud_range,
                cnnseg_feature_dim=6,
                compute_logodds_cfg=None):
        """
        compute_logodds_cfg = {
            compute_logodds=True,
            vis_range=[-120, -120, -3, 120, 120, 2],
            vis_voxel_size=[0.2083, 0.2083, 1.25],
        }
        """
        self.voxel_size = np.array(voxel_size)
        self.point_cloud_range = np.array(point_cloud_range)
        self.cnnseg_feature_dim = cnnseg_feature_dim
        self.bev_map_feature_dim = cnnseg_feature_dim
        grid_size = (self.point_cloud_range[3:] - self.point_cloud_range[:3]) / self.voxel_size
        self.grid_size = np.round(grid_size).astype(np.int32)
        if compute_logodds_cfg is not None:
            self.compute_logodds = compute_logodds_cfg.get("compute_logodds", False)
            self.vis_range = np.array(compute_logodds_cfg.get("vis_range", 
                                    [-60, -60, -5, 60, 60, 5]), dtype = np.float64)
            self.vis_voxel_size = np.array(compute_logodds_cfg.get('vis_voxel_size', [0.2, 0.2, 10]))
            vxsize = int((self.vis_range[3] - self.vis_range[0]) / self.vis_voxel_size[0])
            vysize = int((self.vis_range[4] - self.vis_range[1]) / self.vis_voxel_size[1])
            vzsize = int((self.vis_range[5] - self.vis_range[2]) / self.vis_voxel_size[2])
            self.bev_map_feature_dim += vzsize
            self.vis_grid_size = np.array([vxsize, vysize, vzsize])
            self.vis_origins = np.array(compute_logodds_cfg.get('vis_origin', [0.0, 0.0, 0.0]))
            self.vis_offset_origins = self.vis_origins - self.vis_range[:3]
            self.lo_occupied = 0.7
            self.lo_free = 0.4
            
        else:
            self.compute_logodds = False

    @paddle.no_grad()     
    def __call__(self, points):
        # points has timestamp [x,y,z,i,t]
        if points.shape[1] == 5:
            cur_points_mask = points[:, -1] == 0
            cur_points = points[cur_points_mask, :]
        else:
            cur_points = points            
        
        bev_map_feature = paddle.zeros((int(self.cnnseg_feature_dim), int(self.grid_size[1]), int(self.grid_size[0])))
        bev_map_feature[0, :, :] = -5.
        # calculate points2grid
        pos_xf = (cur_points[:, 0] - self.point_cloud_range[0]) / float(self.voxel_size[0])
        pos_yf = (cur_points[:, 1] - self.point_cloud_range[1]) / float(self.voxel_size[1])
        pos_zf = (cur_points[:, 2] - self.point_cloud_range[2]) / float(self.voxel_size[2])
        
        points2grid = pos_yf.cast("int32") * self.grid_size[1] + pos_xf.cast("int32") 
        mask = (pos_xf < 0) | (pos_xf >= self.grid_size[0])
        mask |= ((pos_yf < 0) | (pos_yf >= self.grid_size[1]))
        mask |= ((cur_points[:, 2] < self.point_cloud_range[2]) | (cur_points[:, 2] > self.point_cloud_range[5]))
        points2grid = paddle.where(mask, paddle.full(points2grid.shape, -1, dtype = 'int32'), points2grid)
        points2grid = points2grid.cast("int32") 
        bev_map_feature = bev_feature_layer.cnnseg_feature_generator_gpu(cur_points, 
                                                            points2grid, 
                                                            self.cnnseg_feature_dim,
                                                            self.grid_size[0],
                                                            self.grid_size[1],
                                                            self.grid_size[2])
        # 0: max_height 1: mean_height 2: count 3: top_intensity 4: mean_intensity
        # 5: nonempty
        bev_map_feature = bev_map_feature.reshape((self.cnnseg_feature_dim, -1))
        nonempty_mask = bev_map_feature[2, ...]  > 0 
        
        bev_map_feature_tmp0 = paddle.where(~nonempty_mask, paddle.zeros(bev_map_feature[0].shape), bev_map_feature[0])
        bev_map_feature_tmp1 = paddle.where(nonempty_mask, (bev_map_feature[1] / (bev_map_feature[2] + 1e-8)), bev_map_feature[1])
        bev_map_feature_tmp4 = paddle.where(nonempty_mask, (bev_map_feature[4] / (bev_map_feature[2] + 1e-8)), bev_map_feature[4])
        bev_map_feature_tmp5 = paddle.where(nonempty_mask, paddle.ones(bev_map_feature[5].shape), bev_map_feature[5])

        bev_map_feature = bev_map_feature.put_along_axis(paddle.to_tensor([[0]]), bev_map_feature_tmp0, axis=0)
        bev_map_feature = bev_map_feature.put_along_axis(paddle.to_tensor([[1]]), bev_map_feature_tmp1, axis=0)
        bev_map_feature = bev_map_feature.put_along_axis(paddle.to_tensor([[4]]), bev_map_feature_tmp4, axis=0)
        bev_map_feature = bev_map_feature.put_along_axis(paddle.to_tensor([[5]]), bev_map_feature_tmp5, axis=0)

        bev_map_feature = bev_map_feature.put_along_axis(paddle.to_tensor([[2]]), 
                                        paddle.log(bev_map_feature[2, ...] + 1), axis=0)
        bev_map_feature = bev_map_feature.reshape((self.cnnseg_feature_dim, self.grid_size[1], self.grid_size[0]))
        if self.compute_logodds:
            logodds = paddle.zeros((int(self.vis_grid_size[2]), int(self.vis_grid_size[1]), int(self.vis_grid_size[0])))
            
            logodds = bev_feature_layer.visibility_feature_gpu(cur_points,
                                                        self.vis_voxel_size[0],
                                                        self.vis_voxel_size[1],
                                                        self.vis_voxel_size[2],
                                                        self.vis_range[0],
                                                        self.vis_range[1],
                                                        self.vis_range[2],
                                                        self.vis_grid_size[0],
                                                        self.vis_grid_size[1],
                                                        self.vis_grid_size[2],
                                                        self.vis_offset_origins[0],
                                                        self.vis_offset_origins[1],
                                                        self.vis_offset_origins[2],
                                                        self.lo_occupied,
                                                        self.lo_free)
            logodds = paddle.where(logodds <= 0, paddle.full(logodds.shape, 0.5, 'float32'), logodds)

            bev_map_feature = paddle.concat([bev_map_feature, logodds], axis=0)
        bev_map_feature = bev_map_feature.reshape((self.bev_map_feature_dim, self.grid_size[1], self.grid_size[0]))
        return bev_map_feature
